### PR TITLE
fix issue120 and issue103: add mutex to prevent data race in websocket write message and use named return values for WritePkg method

### LINKED ISSUE #120: Concurrency issues can occur when using this framework for websocket transfers
WARNING: DATA RACE
Read at 0x00c0004610f8 by goroutine 14:
  github.com/gorilla/websocket.(*Conn).beginMessage()
      /Users/vendor/github.com/gorilla/websocket/conn.go:472 +0x65
  github.com/gorilla/websocket.(*Conn).WriteMessage()
      /Users/vendor/github.com/gorilla/websocket/conn.go:757 +0xf9
  github.com/apache/dubbo-getty.(*gettyWSConn).writePing()
      /Users/vendor/github.com/apache/dubbo-getty/connection.go:627 +0x68
  github.com/apache/dubbo-getty.heartbeat.func1()
      /Users/vendor/github.com/apache/dubbo-getty/session.go:519 +0x6d
  github.com/apache/dubbo-getty.heartbeat()
      /Users/vendor/github.com/apache/dubbo-getty/session.go:533 +0x167
  github.com/dubbogo/gost/time.(*TimerWheel).run()
      /Users/vendor/github.com/dubbogo/gost/time/timer.go:281 +0x2b9
  github.com/dubbogo/gost/time.NewTimerWheel.func1()
      /Users/vendor/github.com/dubbogo/gost/time/timer.go:216 +0x4a4

Previous write at 0x00c0004610f8 by goroutine 7976:
  github.com/gorilla/websocket.(*messageWriter).endMessage()
      /Users/vendor/github.com/gorilla/websocket/conn.go:539 +0xcd
  github.com/gorilla/websocket.(*messageWriter).flushFrame()
      /Users/vendor/github.com/gorilla/websocket/conn.go:626 +0xb85
  github.com/gorilla/websocket.(*Conn).WriteMessage()
      /Users/vendor/github.com/gorilla/websocket/conn.go:763 +0x1d0
  github.com/apache/dubbo-getty.(*gettyWSConn).send()
      /Users/vendor/github.com/apache/dubbo-getty/connection.go:618 +0x117
  github.com/apache/dubbo-getty.(*session).WritePkg()
      /Users/vendor/github.com/apache/dubbo-getty/session.go:408 +0x3a7
  server/common/net.(*Session).Write()

### PR DESCRIPTION
**Pull request by No-SilverBullet**
**What this PR does**:

based on doc: https://pkg.go.dev/github.com/gorilla/websocket#hdr-Concurrency and https://github.com/apache/dubbo-getty/issues/120

1.add lock (sync.Mutex) in gettyWSConn struct
2. add new function 'threadSafeWriteMessage' to ensure that only one thread can send a message at a time,preventing race conditions.

based on https://github.com/apache/dubbo-getty/issues/103
3. use named return values for WritePkg method to return the captured panic to caller, which can prevent the panic  in caller.
**Which issue(s) this PR fixes**:

Fixes https://github.com/apache/dubbo-getty/issues/120, https://github.com/apache/dubbo-getty/issues/103

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Review by FoghostCn — COMMENTED** [review]
Should we use the packetLock in the session?

**Comment by AlexStocks** [comment]
@No-SilverBullet good job. Maybe u can also check the tcp write is thread safe or not. refer issue https://github.com/apache/dubbo-getty/issues/103

**Comment by No-SilverBullet** [comment]
> Should we use the packetLock in the session?

Based on this issue120 and document（ https://pkg.go.dev/github.com/gorilla/websocket#hdr-Concurrency

）, the writeMessage function in gorrila package is not concurrency safe, so a lock is added to prevent race conditions.

**Comment by No-SilverBullet** [comment]
> @No-SilverBullet good job. Maybe u can also check the tcp write is thread safe or not. refer issue #103

ok

**Review by AlexStocks — APPROVED** [review]

**Comment by No-SilverBullet** [comment]
> Should we use the packetLock in the session?

Based on this issue120 and document（ https://pkg.go.dev/github.com/gorilla/websocket#hdr-Concurrency

）, the writeMessage function in gorrila package is not concurrency safe, so a lock is added to prevent race conditions.

**Review by FoghostCn — APPROVED** [review]

File: session.go [hunk 1]
```diff
@@ -381,7 +381,9 @@ func (s *session) WritePkg(pkg interface{}, timeout time.Duration) (int, int, er
 			const size = 64 << 10
 			rBuf := make([]byte, size)
 			rBuf = rBuf[:runtime.Stack(rBuf, false)]
-			log.Errorf("[session.WritePkg] panic session %s: err=%s\n%s", s.sessionToken(), r, rBuf)
+			errMsg := fmt.Sprintf("[session.WritePkg] panic session %s: err=%s\n%s", s.sessionToken(), r, rBuf)
+			log.Error(errMsg)
+			err = perrors.WithStack(fmt.Errorf(errMsg))
```
Comment (Lvnszn) [inline session.go]: 可以先赋值成 error，然后 log 出来

**Review by Lvnszn — APPROVED** [review]

File: session.go (hunk 1 above)
Comment (No-SilverBullet) [reply session.go]: 已经调整

**Pull request merged**